libcxx/include/atomic

```diff
@@ -1939,6 +2019,318 @@ struct atomic<_Tp*>
     _Tp* operator-=(ptrdiff_t __op) _NOEXCEPT          {return fetch_sub(__op) - __op;}
 };
 
+memory_order __atomic_cmpex_fail_reorder(memory_order __o) {
+    if (memory_order_acq_rel == __o)
+        return memory_order_acquire;
+    else if (memory_order_release == __o)
+        return memory_order_relaxed;
+    else
+        return __o;
+}
+
+template <class _Tp, int _Sco = 0>
+struct __atomic_ref {
+private:
+    template <class _Up>
+    using __atomic_ref_enable_binary_op = _EnableIf<
+            is_integral<_Up>::value,
+            bool
+        >;
+
+    template <class _Up>
+    using __atomic_ref_enable_arithmetic = _EnableIf<
+            _Or<
+                is_floating_point<_Up>,
+                is_integral<_Up>
+            >::value,
+            bool
+        >;
+
+    _Tp* __ptr;
+
+public:
+    using value_type = _Tp;
+    using difference_type = value_type;
+
+    static constexpr bool is_always_lock_free = __cxx_is_always_lock_free<_Tp>::__value;
+
+    static constexpr size_t required_alignment = sizeof(_Tp) > alignof(_Tp) ? sizeof(_Tp) : alignof(_Tp);
+
+    static_assert(sizeof(_Tp) == _LIBCUDACXX_ALIGNOF(_Tp), "Size != alignment");
+
+    __atomic_ref() = delete;
+    __atomic_ref& operator=(const __atomic_ref&) = delete;
+
+    _LIBCUDACXX_INLINE_VISIBILITY
+    explicit __atomic_ref(_Tp& __t) : __ptr(&__t) {}
+
+    __atomic_ref(const __atomic_ref&) noexcept = default;
+
+    _LIBCUDACXX_INLINE_VISIBILITY
+    _Tp operator=(_Tp __t) const noexcept
+    {
+        this->store(__t);
+        return __t;
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY
+    operator _Tp() const noexcept { return this->load(); }
+
+    _LIBCUDACXX_INLINE_VISIBILITY
+    bool is_lock_free() const noexcept {
+        return __cxx_atomic_is_lock_free(sizeof(_Tp));
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY
+    void store(_Tp __t, memory_order __m = memory_order_seq_cst) const noexcept {
+        __raw_atomic_store(__ptr, __t, __m, detail::__scope_tag<_Sco>());
```

Comment:
So looking at this (continuing from the previous comment) - does it mean that there's now two layers that the implementer of an atomic backend needs to provide? The `__cxx` layer and the `__raw` layer? Is there a way to require only one of them to be provided?

What I'm thinking here is that there may be a way to just provide a generic `__cxx` layer built on top of `__raw`, call it a new libc++ atomic backend, and have us just provide the `__raw` layer (and maybe the `__cxx` layer types - we could have the definitions of the types and the definitions of the generic layer on top of `__raw` be hidden behind different macros). Then, libcu++'s changes would be:
1) scopes added to all functions and types in this header;
2) types for the `__cxx` layers; and
3) the `__raw` layer.